### PR TITLE
Translate from regex to recognizer combinators

### PR DESCRIPTION
**Pull request by raphlinus**
Gets rid of the dependency on regex and uses recognizer combinators.
The recognizers are just included, rather than having a dependency on
a separate crate, as they're small and simple enough.

Also adds some tests to make sure numeric overflow is handled without
panic (the previous version would not pass these).

### CODE REVIEW

File: src/recognize.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
+// Copyright 2017 Google Inc. All rights reserved.
```
Comment (steveklabnik) [inline src/recognize.rs:1]: how does "all rights reserved" work with the MIT/Apache license here?

**Review by steveklabnik — COMMENTED** [review]
This is amazing raph, thank you so much! I can't say enough good things. I have one question about the license here, but other than that, let's :shipit:

File: src/recognize.rs (hunk 1 above)
Comment (raphlinus) [reply src/recognize.rs:1]: The intent is for it to be covered by the same licenses as the repo. I think the best way to do this (while preserving the Google copyright for the new file) is to just add boilerplate so there's no confusion.

Comment (steveklabnik) [reply src/recognize.rs:1]: Great! I just wanted to make sure.

**Comment by steveklabnik** [comment]
🎊

**Comment by kamalmarhubi** [comment]
@raphlinus thanks so much for this! It's been bugging me for almost a year since I opened a couple of PRs on rustc_version!
